codalab/apps/web/static/js/worksheet/worksheet_interface.jsx

```
@@ -577,6 +577,52 @@ var Worksheet = React.createClass({
         });
     },
 
+    /**
+     * |public_group_id|: UUID of the `public` group
+     * |permission|: permission integer (where 0 = none, 1 = read, 2 = all)
+     **/
+    addPermission: function(group_id, group_name, permission) {
+      var permission_str;
+      if (permission === 0) {
+        permission_str = 'none';
+      } else if (permission === 1) {
+        permission_str = 'read';
+      } else if (permission === 2) {
+        permission_str = 'all';
+      } else {
+        console.error('Invalid permission value: ' + permission);
+      }
+      
+      oldGroupPermissions = this.state.ws.info.group_permissions;
+      var newGroupPermission = {
+        group_name: group_name,
+        group_uuid: group_id,
+        id: 13,
+        permission: permission,
+        permission_str: permission_str
+      };
+      var newGroupPermissions = oldGroupPermissions.concat([newGroupPermission]);
+      this.setState({
+        ws: _.extend({}, this.state.ws, {
+          info: _.extend({}, this.state.ws.info, {
+            group_permissions: newGroupPermissions
+          })
+        })
+      });
+    },
+
+    removePermission: function(publicGroupIndex) {
```

Would it make more sense to just reload the worksheet and let it propagate down?